### PR TITLE
Add --syntax-path cli argument.

### PR DESCRIPTION
**Pull request by CreatorSiSo**
# Changes

This allows loading custom `.sublime-syntax` files via the --syntax-path used for highlighting code in raw elements.

I would love to hear some feedback on whether this is actually wanted in the main branch and if so what I should improve.
(As you can see I added TODO comments in several places)

# Arguments against this change

- Not available for users of the web app

# Possible alternative

The raw function could take a parameter like `additional_syntaxes: array<path>` which probably would have to be cached.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -102,6 +104,25 @@ pub trait World {
...
+    /// The syntec syntax definitions.
+    fn syntax_set(&self) -> &typst::SyntaxSetWrapper {
+        &DEFAULT_SYNTAXES
+    }
+
     /// Try to access a file at a path.
     fn file(&self, path: &Path) -> FileResult<Buffer>;
 }
+
+/// The default syntect syntax definitions.
```
Comment (CreatorSiSo) [inline src/lib.rs:116]: This should propbaly be moved to some other place right?

File: src/lib.rs [hunk 2]
```diff
@@ -102,6 +104,25 @@ pub trait World {
...
+});
+
+pub struct SyntaxSetWrapper(pub SyntaxSet);
+
+impl std::hash::Hash for SyntaxSetWrapper {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        // TODO: This is a really bad hash function
+        state.write_usize(self.0.syntaxes().len());
+    }
+}
```
Comment (CreatorSiSo) [inline src/lib.rs:128]: This should propbaly be moved to some other place right?

File: src/eval/library.rs [hunk 3]
```diff
@@ -59,7 +59,7 @@ pub struct LangItems {
     /// Raw text with optional syntax highlighting: `` `...` ``.
     pub raw: fn(text: EcoString, tag: Option<EcoString>, block: bool) -> Content,
     /// The language names and tags supported by raw text.
-    pub raw_languages: fn() -> Vec<(&'static str, Vec<&'static str>)>,
+    pub raw_languages: fn(world: &dyn World) -> Vec<(EcoString, Vec<EcoString>)>,
```
Comment (CreatorSiSo) [inline src/eval/library.rs:62]: Is it ok to use EcoString here?

**Comment by laurmaedje** [comment]
I am unsure about this. It would be nicer if this could be somehow specified in the Typst file. Then you can use the normal build command and also, for instance, upload grammars to the web app.

**Comment by laurmaedje** [comment]
Something like `#set raw(syntaxes: ("path.sublime-syntax", ..))` could work.

**Comment by laurmaedje** [comment]
Closing for now. Something like this could be useful, but requires more discussion and design work.

**Pull request closed without merging**